fahrschein/src/main/java/org/zalando/fahrschein/domain/Metadata.java

```diff
@@ -83,6 +84,19 @@ public String getPublishedBy() {
         return publishedBy;
     }
 
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Metadata metadata = (Metadata) o;
+        return Objects.equals(eventType, metadata.eventType) && Objects.equals(eid, metadata.eid) && Objects.equals(partition, metadata.partition) && Objects.equals(version, metadata.version) && Objects.equals(publishedBy, metadata.publishedBy) && Objects.equals(occurredAt, metadata.occurredAt) && Objects.equals(receivedAt, metadata.receivedAt) && Objects.equals(flowId, metadata.flowId) && Objects.equals(spanCtx, metadata.spanCtx);
```

Comment:
thanks, I just checked - can you add spanCtx to toString as well?

Reply:
Sure, have just done it.

Reply:
hm.. re-reading the PR, do we have a simpler way to test equality? Shouldn't it be enough to use eid (or to be overly-correct eid & eventType & partition)?

Reply:
oh, learned something:

> A useful technique to detect and handle duplicate events on consumer side is to be idempotent and to check eid field of event metadata. Note: eid checking is not possible using the “undefined” category, as it’s only supplied in the “business” and “data” categories.

https://nakadi.docs.zalando.net/nakadi-event-broker/subscriptions/#client-rebalancing

Reply:
on the other hand "undefined" category event types don't even have metadata, so can safely be ignored.

Reply:
Are you suggesting to limit the `equals` implementation only to the `eid` field?

Reply:
yes, I mean from a business perspective two events are the same if their event-id is the same - to reduce (theoretical) collisions we can of course add a bit more, so I would say "eid & eventType & partition & version" - occuredAt, receivedAt, flowId etc should be immutable anyway.

Reply:
I think it's a good idea to include all attributes in the hashCode and equals methods. This way, the library can see two Metadata objects as the same when all their attributes match, even if they have different business aspects. It should be up to users to decide if two events are equal or not based on their needs (there might be cases where users want to include certain attributes of events in additional metadata attributes ). They can do this by customizing the hashCode and equals methods in their own event objects.

Reply:
Let's not forget we're still talking about Nakadi. Let's assume this piece of code is used in a hot-path of the application, storing events in a hashmap. Why should we consider other attributes than the obvious ones for identity, wasting CPU cycles for no reason? event-id is a uuid (not sure of which uuid-type), but it's already unique besides some theoretical hash collisions.

Reply:
Regarding event id: This is provided by the producer, therefore you may only draw conclusions within one and the same event type. OTOH this is exactly what the producer considered as "identity" within this stream, therefore my statement holds: identity should be event-type + event-id. (version and partition are actually irrelevant here as well)

To my understanding @fixxer's usecase is equals in a test case, so as long as the test is creating metadata with  eventids following the correct semantics, it should be fine right?

Reply:
I appreciate your perspective on this matter. Nakadi doesn't provide a guarantee of uniqueness for event IDs, and this aspect largely depends on the contract in place between producers and consumers. For example, if a producer publishes a batch of events with identical event IDs, Fahrschein should exercise caution in assuming that their metadata and, consequently, the events themselves are identical. This is particularly relevant because events could belong to different partitions and possess varying attributes based on the partition they are associated with. To ensure accuracy, Fahrschein should only consider two metadata objects as equal when all of their attributes match. It's essential that the determination of event equality aligns with the contract established between the producer and consumer, and Fahrschein should refrain from making any unwarranted assumptions regarding event equality.@fixxer concern appears to revolve around maintaining consistent metadata equality when comparing two events, and that's certainly a valid consideration.

Dehora nakadi-java also follows the same approach: https://github.com/dehora/nakadi-java/blob/f3a18c12ba5348de5c66a9ffad3ddb9141296ca3/nakadi-java-client/src/main/java/nakadi/EventMetadata.java#L254

Reply:
hm I get a feeling we're bikeshedding around this topic.  I am willing to "disagree and commit" in this matter.